File: bot/providers/actions.py
from abc import abstractmethod

import numpy as np
import cv2
import time


class Actions(object):
    """ Defines Actions that must be implemented for each Provider"""
    last_img = None

    def __init__(self):
        pass

    @abstractmethod
    def tap(self, x, y):
        raise NotImplementedError("tap not defined")

    @abstractmethod
    def swipe(self, x1, y1, x2, y2):
        raise NotImplementedError("swipe not defined")

    @abstractmethod
    def swipe_time(self, x1, y1, x2, y2, time_amount):
        raise NotImplementedError("swipe time not defined")

    @abstractmethod
    def key_escape(self):
        """Optional Key Escape"""
        raise NotImplementedError("key_escape not defined")

    @abstractmethod
    def take_png_screenshot(self):
        raise NotImplementedError("take_png_screenshot not defined")

    def get_img_from_screen_shot(self, last_one=False, fail=0):
        if last_one and self.last_img is not None:
            return self.last_img
        screen_shot = self.take_png_screenshot()
        nparr = np.fromstring(screen_shot, np.uint8)
        img = cv2.imdecode(nparr, cv2.IMREAD_COLOR)
        if fail == 5:
            raise Exception("Cannot obtain proper image for provider")
        if img is None:
            return self.get_img_from_screen_shot(fail=fail+1)
        self.last_img = img
        return img

    @abstractmethod
    def swipe_right(self, time_sleep=0):
        raise NotImplementedError("swipe right not defined")

    def tapnsleep(self, point, time_sleep, relative=False):
        x, y = point
        self.tap(x, y)
        self.wait_for_ui(time_sleep)

    @abstractmethod
    def wait_for_ui(self, time_sleep):
        raise NotImplementedError("wait_for_ui not defined")

File: bot/providers/actions.go
package providers

import (
	"gocv.io/x/gocv"
	"github.com/patrickmn/go-cache"
	"github.com/adelowo/onecache"
	"time"
	"fmt"
)

var store onecache.Store

const (
	LAST_IMAGE = "LAST-IMAGE"
)

type Actions interface {
	Tap(x, y int)
	Swipe(x1, y1, x2, y2 int)
	SwipeTime(x1, y1, x2, y2, timeAmount int)
	SwipeRight(timeSleep int)
	WaitForUi(timeSleep int)
	TakePNGScreenShot() ([]byte, error)
	GetImgFromScreenShot(fromCache bool, fail int) gocv.Mat
}

type BaseActions struct {
	cache    *cache.Cache
	options  *Options
	provider Provider
}

func (action *BaseActions) TakePNGScreenShot() ([]byte, error) {
	fmt.Println(action.cache)
	panic("error")
}

func (action *BaseActions) Tap(x, y int) {
	panic("implement me")
}

func (action *BaseActions) Swipe(x1, y1, x2, y2 int) {
	panic("implement me")
}

func (action *BaseActions) SwipeTime(x1, y1, x2, y2, timeAmount int) {
	panic("implement me")
}

func (action *BaseActions) SwipeRight(timeSleep int) {
	panic("implement me")
}

func (action *BaseActions) WaitForUi(timeSleep int) {
	time.Sleep(time.Duration(int64(timeSleep)) * time.Second)
}

func (action *BaseActions) GetImgFromScreenShot(fromCache bool, fail int) gocv.Mat {
	if lastImage, found := action.cache.Get(LAST_IMAGE); found && fromCache {
		return lastImage.(gocv.Mat)
	}
	for timesFailed := 0; timesFailed < fail; timesFailed++ {
		// I am cheating here really not a good design pattern
		// Since go will call BaseActions implementation of TakePNGScreenShot
		// When the provider method is not implemented again
		// This way no rewrites are necessary but again bad, so will probably change
		screenShot, err := action.provider.TakePNGScreenShot()
		if err != nil {
			continue
		}
		img := gocv.IMDecode(screenShot, gocv.IMReadColor)
		if img.Empty() {
			continue
		}
		action.cache.Set(LAST_IMAGE, img, cache.DefaultExpiration)
		return img
	}
	log.Fatal("Cannot obtain proper image for provider")
	return gocv.Mat{}
}

func NewActions(o *Options) Actions {
	cache := cache.New(5*time.Minute, 10*time.Minute)
	action := new(BaseActions)
	action.cache = cache
	action.options = o
	action.provider = o.Provider
	return action
}
